src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    pure nothrow @safe @nogc scope
+    bool isInitialized(this Q)() const
+    {
+        return rc !is null;
+    }
+
+    version (CoreUnittest)
+    {
+        pure nothrow @nogc @trusted scope
+        private bool isValueEq(this Q)(uint val) const
+        {
+            return *getUnsafeValue == val;
+        }
+    }
+
+    /**
+    Return a raw pointer to the underlying reference count pointer.
+    This is unsafe and may lead to dangling pointers to invalid memory.
+
+    Returns:
+         A raw pointer to the reference count.
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    pure nothrow @nogc @system
+    const(CounterType*) getUnsafeValue(this Q)() const
+    {
+        return cast(CounterType*) rc;
+    }
+
+    /**
+    A factory function that creates and returns a new instance of a qualified
+    `__RefCount`.
+
+    Params:
+         QualifiedRefCount = a template parameter that is a qualified
+         `__RefCount` type.
+
+    Returns:
+         A new instance of `QualifiedRefCount`.
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    static pure nothrow @nogc @safe
+    auto make(QualifiedRefCount)()
+    {
+        return QualifiedRefCount(1);
+    }
+}
+
+///
+@system @nogc nothrow
+unittest
+{
+    import core.stdc.stdlib : free, malloc;
+
+    struct rcarray
+    {
+        @nogc nothrow:
+
+        private __RefCount rc;
+        int[] payload;
+
+        this(int sz)
+        {
+            rc = __RefCount.make!__RefCount();
+            payload = (cast(int*) malloc(sz * int.sizeof))[0 .. sz];
+        }
+
+        void opAssign(ref rcarray rhs)
+        {
+            if (rc.isUnique)
+            {
+                // If this was the last reference to the payload,
+                // we can safely free
+                free(payload.ptr);
+            }
+
+            // This will update the reference count
+            rc = rhs.rc;
+            // Update the payload
+            payload = rhs.payload;
+        }
+
+        /* Implement copy constructors */
+        this(return scope ref typeof(this) rhs)
+        {
+            rc = rhs.rc;
+            payload = rhs.payload;
+        }
+
+        ~this()
+        {
+            if (rc.isUnique)
+            {
+                // If this was the last reference to the payload,
+                // we can safely free
+                free(payload.ptr);
+            }
+        }
+    }
+
+    auto a = rcarray(42);
+    assert(a.rc.isUnique);
+    {
+        auto a2 = a; // Construct a2 by copy construction
+        assert(!a.rc.isUnique);
+        auto a3 = rcarray(4242);
+        a2 = a3; // Assign a3 into a2; a's ref count drops
+        assert(a.rc.isUnique);
+        a3 = a; // Assign a into a3; a's ref count increases
+        assert(!a.rc.isUnique);
+        // a2 and a3 go out of scope here
+        // a2 is the last ref to rcarray(4242) -> gets freed
+    }
+    assert(a.rc.isUnique);
+}
```

Comment:
Can we add a few more test to increase the coverage of this module and ensure it works as expected?
Maybe even one with multiple threads to ensure the atomic ops work too.

Reply:
> Can we add a few more test to increase the coverage of this module and ensure it works as expected?

A bunch of tests were deleted with [this commit](https://github.com/dlang/druntime/pull/2679/commits/38d8ecf5248f4863fc2e94f8dea286d57830b400) 

> Maybe even one with multiple threads to ensure the atomic ops work too.

There was one which got deleted as well.